ls1intum/Artemis
# `Programming exercises`: Cache dependencies in the Java blackbox template

### Checklist
#### General



#### Server


#### Client
No Changes at the client


#### Changes affecting Programming Exercises
- [x] ~~**High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).~~ Only applicable to the GitLab/Jenkins template. The Aeolus/LocalCI one does not support setting Docker run flags specific to the exercise.


### Motivation and Context

The actual caching approach with Docker volumes has one mature disadvange.
It needs to have a testing run for the solution before a testing run for a student submission
on each host, which is running the containers.
That is a problem if you need more hosts to run the container because of scaling issues.

The new approach uses a dedicated cache server (Maven proxy) that does not have this limitation.

### Description


1. Switch the blackbox Dejagnu template to cache Maven dependencies with Sonatype Nexus.
2. Document how to use the Sonatype Nexus Caching for Maven and Gradle.
3. Document how to restrict / disable the network access for student submission corresponding to the used caching approach.


### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- A course where the student is enrolled / can enrol

1. Log in to Artemis with the Instructor account
    1. Navigate to a course
    2. Create a new Programming Exercise there
         1.  Choose language `Java` 
         2.  Choose type `DejaGnu`
    3. Check if the tests for the solution repository are 100 %
2. Log in to Artemis with the Student account
     1.  Go the new Programming Exercise
     2. Make a submission that should pass a test (or all) -- You could use the contents from the solution repository.

The updated documentation can be found at https://artemis-platform--8351.org.readthedocs.build/en/8351/admin/setup/programming-exercises.html#caching-maven-dependencies-with-sonatype-nexus

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage
Only trivial change in java code.
